template/rust-http/Dockerfile

```diff
@@ -1,32 +1,43 @@
-FROM openfaas/of-watchdog:0.7.2 as watchdog
-
-FROM rust:1.37-alpine as builder
+FROM ghcr.io/openfaas/of-watchdog:0.9.11 as watchdog
+FROM rust:1.56-alpine as build
 
 WORKDIR /home/rust
 
 # Copy all the sources
-COPY function ./function
 COPY main ./main
+COPY function ./function
 
-RUN cd main && cargo build --target x86_64-unknown-linux-musl --release
+RUN --mount=type=cache,target=/usr/local/cargo/registry \
+    cd main && cargo build --target x86_64-unknown-linux-musl --release
 
-FROM alpine:3.11 as runner
+FROM alpine:3.17.2 as ship
+
+COPY --from=watchdog /fwatchdog /usr/bin/fwatchdog
+RUN chmod +x /usr/bin/fwatchdog
 
+ARG ADDITIONAL_PKG=""
 # Install packages and add non-root user
-RUN apk --no-cache add curl ca-certificates \
+RUN apk --no-cache add curl ca-certificates libgcc gcompat ${ADDITIONAL_PKG} \
     && addgroup -S app && adduser -S -g app app
 
-ENV USER=app
+# Split instructions so that buildkit can run & cache
+# the previous command ahead of time.
+RUN mkdir -p /home/app \
+    && chown app /home/app
 
-# Copy of-watchdog binary
-COPY --from=watchdog /fwatchdog /usr/bin/fwatchdog
-RUN chmod +x /usr/bin/fwatchdog
+WORKDIR /home/app
+
+COPY --from=build --chown=app:app /home/rust/main/target/release/main .
+RUN chmod +x . && ldd main # test if required links are available
```

Comment:
Can you explain these changes? Why are they needed now when they weren't before?

Reply:
If I am not mistaken, the build artifact was not available in the earlier iteration.